src/selection/Selection.js

```diff
@@ -52,18 +52,96 @@ define(['EventEmitter'], function (EventEmitter) {
         });
     };
 
+    /**
+     * Clears the selection.
+     */
     Selection.prototype.clear = function () {
         this.selected = [];
-        this.emit('change');
+        this.emit('change', this.selected);
     };
 
     Selection.prototype.primary = function () {
         return this.selected[this.selected.length - 1];
     };
 
-    Selection.prototype.all = function () {
+    Selection.prototype.get = function () {
         return this.selected;
     };
 
+    /**
+     * Selects the selectable object and emits the 'change' event.
+     *
+     * @param {object} selectable an object with element and context properties
+     */
+    Selection.prototype.select = function (selectable) {
+        if (!Array.isArray(selectable)) {
+            selectable = [selectable];
+        }
+
+        if (this.selected[0] && this.selected[0].element) {
+            this.selected[0].element.classList.remove('s-selected');
+        }
+
+        if (this.selected[1]) {
+            this.selected[1].element.classList.remove('s-selected-parent');
+        }
+
+        if (selectable[0] && selectable[0].element) {
+            selectable[0].element.classList.add('s-selected');
+        }
+
+        if (selectable[1]) {
+            selectable[1].element.classList.add('s-selected-parent');
+        }
+
+        this.selected = selectable;
+        this.emit('change', this.selected);
+    };
+
+    Selection.prototype.capture = function (selectable) {
+        if (!this.capturing) {
+            this.capturing = [];
+        }
+
+        this.capturing.push(selectable);
+    };
+
+    Selection.prototype.selectCapture = function (selectable) {
+        if (!this.capturing) {
+            return;
+        }
+
+        this.select(this.capturing.reverse());
+        delete this.capturing;
+    };
+
+    /**
+     * Attaches the click handlers to the element.
+     *
+     * @param element an html element
+     * @param context object with oldItem, item and toolbar properties
+     */
+    Selection.prototype.selectable = function (element, context, select) {
+        var selectable = {
+            context: context,
+            element: element
+        };
+        element.addEventListener('click', this.capture.bind(this, selectable), true);
+        element.addEventListener('click', this.selectCapture.bind(this, selectable));
+        if (select) {
+            this.select(selectable);
+        }
+    };
+
+    /**
+     * Removes the click handlers from the element.
+     *
+     * @param element an html element
+     */
+    Selection.prototype.removeSelectable = function (element) {
+        element.removeEventListener('click', this.capture);
+        element.removeEventListener('click', this.selectCapture);
```

Comment:
These will be different function instances from what was passed to `addEventListener` because of the use of `bind`, so I believe this will cause listener leak. A pattern we've used elsewhere to avoid this is to return a function like `removeSelectable` from `selectable`, e.g.:

```
Selection.prototype.selectable = function (element, context, select) {
        var selectable = {
            context: context,
            element: element
        };
        var capture = this.capture.bind(this, selectable);
        var selectCapture = this.selectCapture.bind(this, selectable);
        element.addEventListener('click', capture, true);
        element.addEventListener('click', selectCapture);
        if (select) {
            this.select(selectable);
        }
        return function () {
            element.removeEventListener('click', capture);
            element.removeEventListener('click', selectCapture);
        };
    };
};
```

The downside of this approach is that it's a little harder to document.

Reply:
Done